clean_build.sh

```diff
@@ -1,32 +1,74 @@
 #!/bin/sh
+
 cd "$(dirname "$0")" || exit 1
+
 # some environments have cmake v2 as 'cmake' and v3 as 'cmake3'
 # check for cmake3 first then fallback to just cmake
-B_CMAKE=`type cmake3 2>/dev/null`
-if [ $? -eq 0 ]; then
-    B_CMAKE=`echo "$B_CMAKE" | cut -d' ' -f3`
+
+B_CMAKE=$(command -v cmake3 2>/dev/null)
+if [ "$?" -eq 0 ]; then
+    # Continue, cmake3 exists in $PATH.
+    continue
 else
-    B_CMAKE=cmake
+    # OK, so cmake3 isn't in path, so let's test to see if `cmake` itself exists, before proceeding.
+    if command -v cmake 2>/dev/null; then
+        B_CMAKE=$(command -v cmake)
+        # We have a cmake executable available, now let's proceed!
+    else
+        # As self-explanatory, the cmake executable isn't available, so we should fail here.
+        echo "ERROR: CMake not in $PATH, cannot build! Please install CMake, or if this persists, file a bug report."
+        exit 1
+    fi
 fi
-# default build configuration
+
+# Set default build type, but in CI, this is set to Release.
+# For downstream distributions, our recommendation is to set this type in your build scripts to Release, and supply Debug build types in your debug packages.
+
 B_BUILD_TYPE=${B_BUILD_TYPE:-Debug}
+
 if [ "$(uname)" = "Darwin" ]; then
-    # OSX needs a lot of extra help, poor thing
-    # run the osx_environment.sh script to fix paths
-    . ./osx_environment.sh
-    B_CMAKE_FLAGS="-DCMAKE_OSX_SYSROOT=$(xcode-select --print-path)/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -DCMAKE_OSX_DEPLOYMENT_TARGET=10.9 $B_CMAKE_FLAGS"
+    # macOS needs a little help, so we source this environment script to fix paths.
+    if [ ! -e "./macOS_environment.sh" ]; then
+        echo "macOS environment script not found, this isn't meant to happen!"
+        exit 1
+    else
+        . ./macOS_environment.sh
+    fi
+
+    B_CMAKE_FLAGS="-DCMAKE_OSX_SYSROOT=$(xcode-select --print-path)/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -DCMAKE_OSX_DEPLOYMENT_TARGET=10.9 ${B_CMAKE_FLAGS:-}"
+fi
+
+# Source local build variables to the environment, if available.
+# If not, continue as normal, and silently.
+if [ -e "./build_env.sh" ]; then
+    . "./build_env.sh"
 fi
-# allow local customizations to build environment
-[ -r ./build_env.sh ] && . ./build_env.sh
 
 # Initialise Git submodules
 git submodule update --init --recursive
 
-B_CMAKE_FLAGS="-DCMAKE_BUILD_TYPE=$B_BUILD_TYPE $B_CMAKE_FLAGS"
-rm -rf build
-mkdir build || exit 1
-cd build || exit 1
-echo "Starting Barrier $B_BUILD_TYPE build..."
-$B_CMAKE $B_CMAKE_FLAGS .. || exit 1
-make || exit 1
-echo "Build completed successfully"
+B_CMAKE_FLAGS="-DCMAKE_BUILD_TYPE=${B_BUILD_TYPE} ${B_CMAKE_FLAGS:-}"
```

Comment:
Having default B_CMAKE_FLAGS would reduce noise here

Reply:
Simplified and fixed in 5e7fa733 - thoughts?

Reply:
Just a reminder - I'm all ready to merge if you are, I've rebased to master as well.